metadata-ingestion/src/datahub/ingestion/source/hex/api.py

```diff
@@ -0,0 +1,178 @@
+
+    def _auth_header(self):
+        return {"Authorization": f"Bearer {self.token}"}
+
+    def fetch_projects(
+        self,
+        include_components: bool = True,
+        include_archived: bool = False,
+        include_trashed: bool = False,
+    ) -> Generator[Union[Project, Component], None, None]:
+        """Fetch all projects and components
+
+        https://learn.hex.tech/docs/api/api-reference#operation/ListProjects
+        """
+        params = {
+            "includeComponents": include_components,
+            "includeArchived": include_archived,
+            "includeTrashed": include_trashed,
+            "includeSharing": True,
+            "limit": self.page_size,
+            "after": None,
+            "before": None,
+            "sortBy": "CREATED_AT",
+            "sortDirection": "ASC",
+        }
+        yield from self._fetch_projects_page(params)
+
+        while params["after"]:
+            yield from self._fetch_projects_page(params)
+
+    def _fetch_projects_page(
+        self, params: Dict[str, Any]
+    ) -> Generator[Union[Project, Component], None, None]:
+        logger.debug(f"Fetching projects page with params: {params}")
+        self.report.fetch_projects_page_calls += 1
+        response = requests.get(
+            url=self._list_projects_url(),
+            headers=self._auth_header(),
+            params=params,
+        )
+        response.raise_for_status()
+        data = response.json()
+        logger.debug(f"Fetched data: {data}")
+        values = data.get("values", [])
+        params["after"] = data.get("pagination", {}).get("after")
+
+        self.report.fetch_projects_page_items += len(values)
+        for value in values:
+            yield self._map_data(data=value)
+
+    def _map_data(self, data: Dict[str, Any]) -> Union[Project, Component]:
+        description = data.get("description")
+        created_at = (
+            self._parse_datetime(data["createdAt"]) if data.get("createdAt") else None
+        )
+        last_edited_at = (
+            self._parse_datetime(data["lastEditedAt"])
+            if data.get("lastEditedAt")
+            else None
+        )
+        status = Status(name=data["status"]["name"]) if data.get("status") else None
+        categories = [
+            Category(
+                name=cat["name"],
+                description=cat.get("description"),
+            )
+            for cat in data.get("categories", [])
+        ]
+        collections = [
+            Collection(
+                name=col["collection"]["name"],
+            )
+            for col in data.get("sharing", {}).get("collections", [])
+        ]
+        creator = Owner(email=data["creator"]["email"]) if data.get("creator") else None
+        owner = Owner(email=data["owner"]["email"]) if data.get("owner") else None
+        analytics = (
+            Analytics(
+                appviews_all_time=data["analytics"]["appViews"].get("allTime"),
```

Comment:
is there any way we can leverage pydantic to simplify/eliminate all this parsing logic?

Reply:
I did in the commit 7514d23602fca33027f60248c72b8d3c6ede53dc

I wanted to decouple the domain model from the api model, so the cost is to have two models. No prob.

The good thing is, as you suggested, pydantic manages the deserialization, and model constraints are explicit in the model, rather that implicit in the code.